dispatcher.go

```diff
@@ -103,16 +109,35 @@ func NewDispatcher(cfg Config) *Dispatcher {
 		panic("yarpc.NewDispatcher expects a valid service name: %s" + err.Error())
 	}
 
+	logger := zap.NewNop()
+	if cfg.Logger != nil {
+		logger = cfg.Logger.Named("yarpc").With(
+			zap.Namespace("yarpc"), // isolate yarpc's keys
+			zap.String("dispatcher", cfg.Name),
+		)
+		cfg = addObservingMiddleware(cfg, logger)
+	}
+
 	return &Dispatcher{
 		name:              cfg.Name,
 		table:             middleware.ApplyRouteTable(NewMapRouter(cfg.Name), cfg.RouterMiddleware),
 		inbounds:          cfg.Inbounds,
 		outbounds:         convertOutbounds(cfg.Outbounds, cfg.OutboundMiddleware),
 		transports:        collectTransports(cfg.Inbounds, cfg.Outbounds),
 		inboundMiddleware: cfg.InboundMiddleware,
+		log:               logger,
 	}
 }
 
+func addObservingMiddleware(cfg Config, logger *zap.Logger) Config {
```

Comment:
In the past, we've tried to follow the philosophy of not using our own extension points to provide core functionality. I think logging might fall into that group.

@kriskowal @willhug @abhinav are ya'll fine with using middleware in this way?